Fusion vcf view support

fevac opened the pull request:
This PR adds a functionality or fixes a bug.

Closes https://github.com/Clinical-Genomics/scout/issues/4110

Closes https://github.com/Clinical-Genomics/MTP-RNAFUSION/issues/8

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>


**How to test**:
1. how to test it, possibly with real cases/data

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by CR
- [x] tests executed by EC, CR

northwestwitch commented:
I've maked some false positives reported by SonarCloud as bugs ( now there are 2 instead of 15). I've also fixed the imports so the linting checks are fine. Unfortunately there is an error and the automatic tests don't pass. I'll continue checking this PR tomrorrow

northwestwitch commented:
I was thinking that perhaps we could do the review like this:
- Reach the point where the functionality works as expected without caring too much for the code
- Do the code review

First test:

- Switch to this branch
- Run `scout setup demo` -> Runs OK without errors
- Go to the demo case `rnafusion_case`. There I see that there is the fusion variants button. Unfortunately when I click on it I get an error:

<img width="846" alt="image" src="https://github.com/Clinical-Genomics/scout/assets/28093618/31606816-5a9f-44e7-9d1f-f99d7cc04eda">


```
2023-11-16 07:21:26 n187-p80.local scout.adapter.mongo.query[7270] INFO mongo query: {'case_id': 'solidfawn', 'category': 'fusion', 'variant_type': 'clinical'}
2023-11-16 07:21:26 n187-p80.local scout.server.app[7270] ERROR Exception on /cust000/rnafusion_case/fusion/variants [GET]
Traceback (most recent call last):
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/flask/app.py", line 2529, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/flask/app.py", line 1825, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/flask_cors/extension.py", line 176, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/flask/app.py", line 1823, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/utils.py", line 86, in decorated_function
    context = func(*args, **kwargs)
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/blueprints/variants/views.py", line 662, in fusion_variants
    data = controllers.fusion_variants(
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/blueprints/variants/controllers.py", line 332, in fusion_variants
    parse_variant(
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/blueprints/variants/controllers.py", line 886, in parse_variant
    genes_have_changed = update_variant_genes(store, variant_obj, genome_build)
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/blueprints/variants/controllers.py", line 510, in update_variant_genes
    if not gene_obj["hgnc_id"]:
TypeError: 'NoneType' object is not subscriptable
```

Perhaps I should have loaded the fusion variants via the cli, like scout load variants instead? But I don't see the option in the cli:

Available options:

```
  -i, --institute TEXT            institute id of related cases
  --cancer                        Upload clinical cancer variants
  --cancer-research               Upload research cancer variants
  --cancer-sv                     Upload clinical cancer structural variants
  --cancer-sv-research            Upload research cancer structural variants
  --mei                           Upload clinical MEI variants
  --mei-research                  Upload research MEI variants
  --sv                            Upload clinical structural variants
  --sv-research                   Upload research structural variants
  --snv                           Upload clinical SNV variants
  --snv-research                  Upload research SNV variants
  --str-clinical                  Upload clinical STR variants
  --chrom TEXT                    If region, specify the chromosome
  --start INTEGER                 If region, specify the start
  --end INTEGER                   If region, specify the end
  --hgnc-id INTEGER               If all variants from a gene, specify the
```

northwestwitch commented:
Hi agan @fevac. I've been trying to load the case as you suggested yesterday (kinfof):

I am using scout demo locally. Since scout demo already has a rnafusion case (solidfawn), I need to remove it first:
- `scout --demo delete case -c solidfawn`

Then I load the case as a new one:
- `scout --demo load case scout/demo/rnafusion.load_config.yaml`

The case gets loaded without errors, but the if I go to case page and click in the variants button:
<img width="154" alt="image" src="https://github.com/Clinical-Genomics/scout/assets/28093618/646fd6e7-e775-4637-8c58-0f0d16a2a905">

It crashes exactly with the same error I posted in my comment above.

<hr>

Additionally I've noticed that if you run `scout setup demo` from this branch, and then you try to see cancer variantS (either SNVs or SVs) then the variantS page crashes with this error, which is exactly the error you get in the test that is not passing at the moment!

```
File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/templates/layout.html", line 1, in top-level template code
    {% extends "bootstrap_global.html" %}
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/templates/bootstrap_global.html", line 62, in top-level template code
    {% block body %}
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/templates/bootstrap_global.html", line 65, in block 'body'
    {% block content %}
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/templates/layout.html", line 93, in block 'content'
    {% block content_main %}{% endblock %}
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/blueprints/variants/templates/variants/cancer-variants.html", line 114, in block 'content_main'
    <td>{{ rank_cell(variant) }}</td>
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/jinja2/runtime.py", line 777, in _invoke
    rv = self._func(*arguments)
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/blueprints/variants/templates/variants/cancer-variants.html", line 150, in template
    {% if variant.rank_score <= 4 %}
TypeError: '<=' not supported between instances of 'str' and 'int'

```

northwestwitch commented:
After the latest commits the fusion variants are shown! 💯 

<img width="1439" alt="image" src="https://github.com/Clinical-Genomics/scout/assets/28093618/a6c9f92d-5129-49c3-b143-02e50b9facb9">


I'm going to review the code now!

fevac commented:
Hi @northwestwitch
Demo working now. I was not planning to add `scout  load variants --fusion` as we will upload via the config. But maybe you want me to add it?

northwestwitch commented:
> Hi @northwestwitch Demo working now. I was not planning to add `scout load variants --fusion` as we will upload via the config. But maybe you want me to add it?

I would reccommend doing in in a follow up PR. This looks already huge as it is, no?

northwestwitch commented on scout/parse/variant/variant.py:
```diff
@@ -418,6 +424,94 @@ def set_str_info(variant: Variant, parsed_variant: Dict[str, Any]):
+
+    genes = []
+    for suffix in ["a", "b"]:
+        genes.append(
```
I would have only one for loop `for suffix in ["a", "b"]:`and inside it handle both parsed stranscripts and genes. So you avoid repeated code

northwestwitch commented on scout/server/blueprints/variant/templates/variant/fusion-variant.html:
```diff
@@ -0,0 +1,464 @@
+    </a>
+  </li>
+  <li class="nav-item">
+    {% if variant.category == "fusion" %}
```
But if you are using this template you already know that variant.category == "fusion", no?

northwestwitch commented on scout/server/blueprints/variant/templates/variant/fusion-variant.html:
```diff
@@ -0,0 +1,464 @@
+          Category
+          <span class="badge rounded-pill bg-secondary float-end">{{ variant.sub_category|upper }}</span>
+        </li>
+        {% if variant.category == "fusion" and variant.somatic_score %}
```
Same here, you know that these are fusion variants

northwestwitch commented on scout/server/blueprints/cases/templates/cases/case.html:
```diff
@@ -265,6 +265,9 @@ <h4 class="">
         {% if case.vcf_files.vcf_cancer_sv %}
           <a class="btn btn-dark btn-sm text-white" href="{{ url_for('variants.cancer_sv_variants', institute_id=institute._id, case_name=case.display_name, variant_type='clinical') }}">Clinical {% if case.track=="cancer" %}somatic {% endif %}structural variants</a>
         {% endif %}
+        {% if case.vcf_files.vcf_fusion %}
```
Would be nice to have a direct link from caseS page, that now looks like this:

<img width="806" alt="image" src="https://github.com/Clinical-Genomics/scout/assets/28093618/045ad4d4-7839-4199-a5d4-37d89cd3bc4a">

northwestwitch commented on scout/server/blueprints/variants/templates/variants/fusion-variants.html:
```diff
@@ -0,0 +1,189 @@
+          <tr>
+            <th style="width:3%"></th>
+            <th style="width:4%" title="Beta version. Currently testing">Rank (beta)</th>
+            <th style="width:6%" title="HGNC symbols">Fusion Genes</th>
```
```suggestion
            <th style="width:6%" title="Genes">Fusion Genes</th>
```

northwestwitch commented on scout/server/blueprints/variants/templates/variants/fusion-variants.html:
```diff
@@ -0,0 +1,189 @@
+          {% for gene in variant.genes %}
+             <a data-bs-toggle="tooltip" data-bs-html="true" title="
+               <div>
+                    <strong>{{ gene.hgnc_symbol or gene.hgnc_id }}</strong>:  {{ gene.description }}
```
Ah see, it should display symbol or if symbol is missing then the HGNC ID.

**What I would expect on scout stage: display HGNC ID**, since HGNC symbol is missing in the demo databaae
**What I would expect in a non-demo instance: display symbo**l, since genes are found in the database

northwestwitch commented on scout/build/variant/gene.py:
```diff
@@ -60,6 +61,9 @@ def build_gene(gene, hgncid_to_gene=None):
             gene_obj["inheritance"] = hgnc_gene["inheritance_models"]
         if hgnc_gene.get("phenotypes"):
             gene_obj["phenotypes"] = hgnc_gene["phenotypes"]
+    else:
```
No, I wouldn't have this in the code. The idea is to always start from HGNC IDs to get gene symbols associated to a variant. This is because symbols change over time but IDs are static. Probably you added this code because of demo db not having these genes, but I think you should instead have empty symbols and only use HGNC IDs instead. This happens also when in other demo cases where the genes are not in the database, see for example demo cancer case, where only the first gene is present in the database:

<img width="624" alt="image" src="https://github.com/Clinical-Genomics/scout/assets/28093618/eaecd919-7f45-4a39-931e-9e0810220123">


I know that filling in missing symbols with stuff parsed from the VCF is tempting, but might lead to errors, especially for genea that have been renamed over time

fevac replied:
So basically, this is to overcome the issue with some fusions for which we have gene symbols but no hgnc_ids 
But I could remove this and add those genes in a fusion_genes variable instead

northwestwitch replied:
> So basically, this is to overcome the issue with some fusions for which we have gene symbols but no hgnc_ids
> But I could remove this and add those genes in a fusion_genes variable instead

I don't think you need to add an additional variable here. I would just do like this:

- If you have BOTH HGNC ID and SYMBOL, use HGNC to retrieve symbol
- If you have just SYMBOL use SYMBOL ?

That should take care of not messing up with the other types of variants I think

northwestwitch reviewed (changes requested):
Great work!

This PR adds a lot of features at the same time (which is great) but I think for the sake of testing it properly you could avoid at least the loading of the case with the --vcf-fusion parameters and have it in another PR?

One thing that I would change, as I explained in the comments, is when you collect gene symbols from the VCF instead of going thru HGNC ID in VCF --> Gene in database  --> Gene symbol. Gene symbols change over time and showing just what you see in the VCF could lead to errors.

I have a buch of other suggestions as well but make also sure you pay attention to what is suggested by SonarCloud. 

Last think: you should add a test for fusion variantS and fusion variant web pages. Take a loot at how these tests are done [here](https://github.com/Clinical-Genomics/scout/blob/6fa2237d9d81f8f215ce5818a256fb0eea1068d9/tests/server/blueprints/variants/test_variants_views.py#L67) and [here](https://github.com/Clinical-Genomics/scout/blob/6fa2237d9d81f8f215ce5818a256fb0eea1068d9/tests/server/blueprints/variant/test_variant_views.py#L79)

What's going on with the automatic tests?

fevac commented on scout/build/variant/gene.py:
```diff
@@ -50,6 +50,7 @@ def build_gene(gene, hgncid_to_gene=None):
     inheritance = set()
     hgnc_transcripts = []
+
     if hgnc_gene:
```
We always have the hgnc symbols but not always have the hgnc IDs. So in those cases where there is no hgnc_id we want to keep the symbol

northwestwitch replied:
I understand. But then please check that this PR doesn't affect the parsing of other types of variants. For instance demo cancer variants on main branch looks like this:

<img width="354" alt="image" src="https://github.com/Clinical-Genomics/scout/assets/28093618/f03f88fa-5402-4d30-ab65-503ef697ee7a">


If you run the setup using this branch, then it looks like this:

<img width="392" alt="image" src="https://github.com/Clinical-Genomics/scout/assets/28093618/01a7ada2-898c-4240-808d-b7d15c394414">

fevac commented on scout/parse/variant/genotype.py:
```diff
@@ -182,6 +184,24 @@ def get_mei_reads(variant: cyvcf2.Variant, pos: Dict[str, int]) -> Tuple[int, ..
     return (spanning_ref, clip5_alt, clip3_alt)
 
 
+def get_ffpm_info(variant: cyvcf2.Variant, pos: Dict[str, int]) -> Tuple[int, ...]:
```
oh, yes! I changed this method and this was fully updated. Good catch.

fevac commented on scout/commands/load/case.py:
```diff
@@ -39,6 +39,11 @@
     type=click.Path(exists=True),
     help="path to clinical STR VCF file to be loaded",
 )
+@click.option(
```
reverted

fevac commented on scout/commands/update/case.py:
```diff
@@ -74,10 +74,12 @@ def case(
     vcf_sv,
     vcf_cancer,
     vcf_cancer_sv,
+    vcf_fusion,
```
reverted

fevac commented:
Hi @northwestwitch thanks for the review. I went through your comments now. So, you can go for a second round of review if you want to. 

I will check now the tests you mentioned and try to add some for this.

northwestwitch commented on scout/server/blueprints/variants/templates/variants/fusion-variants.html:
```diff
@@ -0,0 +1,189 @@
+    </td>
+    <td>
+      <div class="align-items-center">
+        {% if variant.genes %}
```
Ah I see that links to genes without variant.genes (where the HGNC ID is missing) are not present:

<img width="393" alt="image" src="https://github.com/Clinical-Genomics/scout/assets/28093618/e565635f-a4b3-4238-ae05-5bb8969f246f">

Just out of curiosity, how come these variants don't have HGNC IDs??

fevac replied:
so the pipeline gives gene symbols and Annick added a step to retrieve hgnc_ids but some IDs they cannot be found, for example `IGH@`

northwestwitch commented on scout/parse/variant/variant.py:
```diff
@@ -418,6 +424,93 @@ def set_str_info(variant: Variant, parsed_variant: Dict[str, Any]):
+def set_fusion_info(variant: Variant, parsed_variant: Dict[str, Any]):
+    """Add Fusion information if present."""
+
+    def replace_nan(value: str, nan_value: str = "nan", replace_by: Any = "") -> str:
```
You might apply the function also to the orientantion:

<img width="279" alt="image" src="https://github.com/Clinical-Genomics/scout/assets/28093618/e2475661-5739-4cab-8f94-e9d9432cbf6f">

northwestwitch commented on scout/parse/variant/variant.py:
```diff
@@ -418,6 +424,93 @@ def set_str_info(variant: Variant, parsed_variant: Dict[str, Any]):
+    parsed_variant["chr_b"] = call_safe(str, variant.INFO.get("CHRB", ""))
+    parsed_variant["position_a"] = call_safe(str, variant.INFO.get("POSA", ""))
+    parsed_variant["position_b"] = call_safe(str, variant.INFO.get("POSB", ""))
+    parsed_variant["breakpoint_a"] = f"{parsed_variant['chr_a']}:{parsed_variant['position_a']}"
```
Check the last variant in the demo variants. It has not much info about the coordinates, and breakpoints 0:0 doesn't make much sense. Perhaps replace 0 with "-" ? 

<img width="130" alt="image" src="https://github.com/Clinical-Genomics/scout/assets/28093618/de0d87d8-e32a-406d-99b9-675ed0fc8ccf">

Also consider adding "-" instead of nothing for the missing positions of variant 2 and 3

fevac replied:
I've realised I can get the break points from the coordinates position and end so I we don't need to store the breakpoint information. I will update this and remove it

fevac replied:
Removed breakpoint parsing in favour of postions, now all fusions should display breakpoints.

northwestwitch reviewed (changes requested):
Amazing job. I feel that the functionality now works as intended. I have a few minor suggestions, see my comments.

As I mentioned, would be nice to add tests for the fusion variants and fusion variant page. As soon as we do this it is good to merge I think. Let me know if you want help writing the tests

fevac commented:
Hi @northwestwitch I added a test for the fusion_variants.
I removed the fusion_variant html at the end as I wasn't using it and kept the general variant view for now. 
Please feel free to make changes in this branch if you want to add more tests. You think we can merge soon?

fevac commented:
there is no variant page. Only variantS

northwestwitch commented:
> there is no variant page. Only variantS

variant page opens when you click on the rank of one of the variants?

fevac commented:
it goes to the general variant.variant page. Not to fusion specific variant page

fevac commented:
but yes, please have look tomorrow I might be misunderstanding

northwestwitch commented:
> it goes to the general variant.variant page. Not to fusion specific variant page

Yes it's the usual variant page, but you are sending to it the data from a new type of variant (the fusion one), which has other key/values compared with the other types of variants the page is already tested with

northwestwitch commented on tests/server/blueprints/variants/test_variants_views.py:
```diff
@@ -254,6 +254,27 @@ def test_str_variants(app, institute_obj, case_obj):
         assert resp.status_code == 200
 
 
+def test_fusion_variants(app, institute_obj, case_obj):
```
Nice with this tests. I see that you are using case_obj (the demo case witch is a WES case) in the testing. I think it would be more appropriate to use the other demo case, the rna fusion case. What happens in this test is that you are passing WES variants to the fusion variants page, not fusion variants. I'll see if it's already available in the test fixtures so we can use it!

northwestwitch commented on tests/conftest.py:
```diff
@@ -1211,6 +1214,37 @@ def sv_variant_objs(request, parsed_sv_variants, institute_obj):
+
+
+@pytest.fixture(scope="function")
+def fusion_variant_obj(request, parsed_fusion_variants, institute_obj) -> dict:
```
I've added a few fixtures that should also indirectly test the parsing of the fusion variants. Something is not working as expected still..

northwestwitch commented:
I'm working on the tests. In the meanwhile we could try to load variants for a case from hasta stage if you have some test data @fevac

northwestwitch reviewed (approved):
Ok, this PR looks good enough now I think. Before merging, don't we have some real data that can be associated to a test case and loaded via Hasta?

fevac commented:
tests are passing! Just a bit unclear what are the NAN values reported by pipeline, so let me confirm before merging

fevac commented:
ok, got confirmation. Nan values should be as they are here, so "nan"

henrikstranneheim commented:
🎉 Well done!

northwestwitch commented:
Case with fusion variants, for future reference & testing: https://scout-stage.scilifelab.se/cust000/ACC8869A1-1M-FUSION
